File: src/routes/vendor-attachment.routes.ts
import { Hono } from "hono";
import type { Env } from "@/types/hono";
import { VendorAttachmentService, type UploadAttachmentInput } from "@/services/vendor-attachment.service";
import { successResponse } from "@/utils/response";
import { authMiddleware } from "@/middleware/auth";
import { z } from "zod";
import { Binary } from "mongodb";

export const vendorAttachmentRoutes = new Hono<Env>();

let service: VendorAttachmentService | null = null;

function getService(): VendorAttachmentService {
  if (!service) {
    service = new VendorAttachmentService();
  }
  return service;
}

// Apply auth middleware
vendorAttachmentRoutes.use("*", authMiddleware);

// GET /api/vendors/:vendorId/attachments - List attachments for vendor (metadata only)
vendorAttachmentRoutes.get("/:vendorId/attachments", async (c) => {
  const vendorId = c.req.param("vendorId");

  const attachments = await getService().listAttachmentsByVendor(vendorId);

  return c.json(successResponse(attachments));
});

// GET /api/vendors/:vendorId/attachments/:attachmentId - Get attachment metadata
vendorAttachmentRoutes.get("/:vendorId/attachments/:attachmentId", async (c) => {
  const attachmentId = c.req.param("attachmentId");

  const attachment = await getService().getAttachmentMetadata(attachmentId);

  return c.json(successResponse(attachment));
});

// GET /api/vendors/:vendorId/attachments/:attachmentId/download - Download PDF
vendorAttachmentRoutes.get("/:vendorId/attachments/:attachmentId/download", async (c) => {
  const attachmentId = c.req.param("attachmentId");

  const attachment = await getService().getAttachment(attachmentId);

  // Return binary data with correct content type
  const buffer = (attachment.data as Binary).buffer;
  return new Response(buffer, {
    headers: {
      "Content-Type": attachment.mimeType,
      "Content-Length": attachment.size.toString(),
      "Content-Disposition": `attachment; filename*=UTF-8''${encodeURIComponent(attachment.filename)}`,
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "GET, OPTIONS",
      "Access-Control-Allow-Headers": "Content-Type, Authorization",
    },
  });
});

// POST /api/vendors/:vendorId/attachments - Upload PDF (sudo only)
vendorAttachmentRoutes.post("/:vendorId/attachments", async (c) => {
  const user = c.get("user");
  const vendorId = c.req.param("vendorId");

  const formData = await c.req.formData();
  const file = formData.get("file") as File;
  const title = formData.get("title") as string;
  const description = formData.get("description") as string || "";

  if (!file) {
    return c.json({ success: false, error: { message: "No file provided" } }, 400);
  }

  if (!title) {
    return c.json({ success: false, error: { message: "Title is required" } }, 400);
  }

  const input: UploadAttachmentInput = {
    title,
    description,
    filename: file.name,
    mimeType: file.type,
    size: file.size,
    data: Buffer.from(await file.arrayBuffer()),
  };

  const attachment = await getService().uploadAttachment(
    vendorId,
    user._id!.toString(),
    user.role,
    input
  );

  return c.json(successResponse(attachment), 201);
});

// PUT /api/vendors/:vendorId/attachments/:attachmentId - Update attachment metadata (sudo only)
vendorAttachmentRoutes.put("/:vendorId/attachments/:attachmentId", async (c) => {
  const user = c.get("user");
  const attachmentId = c.req.param("attachmentId");
  const body = await c.req.json();

  const updateSchema = z.object({
    title: z.string().optional(),
    description: z.string().optional(),
  });

  const input = updateSchema.parse(body);
  const attachment = await getService().updateAttachment(attachmentId, user.role, input);

  return c.json(successResponse(attachment));
});

// DELETE /api/vendors/:vendorId/attachments/:attachmentId - Delete attachment (sudo only)
vendorAttachmentRoutes.delete("/:vendorId/attachments/:attachmentId", async (c) => {
  const user = c.get("user");
  const attachmentId = c.req.param("attachmentId");

  await getService().deleteAttachment(attachmentId, user.role);

  return c.json(successResponse({ message: "Attachment deleted successfully" }));
});

File: fisoft/src/main/kotlin/routes/VendorAttachmentRoutes.kt
package com.ilyasdemirkiran.routes

import com.ilyasdemirkiran.middleware.authenticateUser
import com.ilyasdemirkiran.middleware.currentUser
import com.ilyasdemirkiran.services.VendorAttachmentService
import com.ilyasdemirkiran.utils.successResponse
import io.ktor.http.*
import io.ktor.http.content.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.serialization.Serializable

@Serializable
data class UpdateVendorAttachmentReq(
    val title: String? = null,
    val description: String? = null
)

fun Route.vendorAttachmentRoutes() {
    val attachmentService = VendorAttachmentService()

    route("/api/vendors/{vendorId}/attachments") {
        // GET /api/vendors/{vendorId}/attachments - List attachments for vendor
        get {
            call.authenticateUser()
            val vendorId = call.parameters["vendorId"] ?: ""
            val attachments = attachmentService.listAttachmentsByVendor(vendorId)
            call.respond(HttpStatusCode.OK, successResponse(attachments))
        }

        // POST /api/vendors/{vendorId}/attachments - Upload attachment PDF (sudo only)
        post {
            call.authenticateUser()
            val user = call.currentUser
            val vendorId = call.parameters["vendorId"] ?: ""
            val multipart = call.receiveMultipart()
            var filename = "attachment.pdf"
            var mimeType = "application/pdf"
            var title = ""
            var bytes: ByteArray? = null

            multipart.forEachPart { part ->
                when (part) {
                    is PartData.FileItem -> {
                        filename = part.originalFileName ?: "attachment.pdf"
                        mimeType = part.contentType?.toString() ?: "application/pdf"
                        bytes = part.streamProvider().readBytes()
                    }
                    is PartData.FormItem -> {
                        if (part.name == "title") {
                            title = part.value
                        }
                    }
                    else -> {}
                }
                part.dispose()
            }

            if (bytes == null) {
                call.respond(HttpStatusCode.BadRequest, mapOf("success" to false, "error" to "No file provided"))
                return@post
            }
            if (title.isEmpty()) {
                title = filename
            }

            val doc = attachmentService.uploadAttachment(user.role, vendorId, title, filename, bytes!!, mimeType)
            call.respond(HttpStatusCode.Created, successResponse(doc))
        }

        // GET /api/vendors/{vendorId}/attachments/{attachmentId} - Get metadata
        get("/{attachmentId}") {
            call.authenticateUser()
            val attachmentId = call.parameters["attachmentId"] ?: ""
            val doc = attachmentService.getAttachmentMetadata(attachmentId)
            call.respond(HttpStatusCode.OK, successResponse(doc))
        }

        // GET /api/vendors/{vendorId}/attachments/{attachmentId}/preview - Inline preview
        get("/{attachmentId}/preview") {
            call.authenticateUser()
            val attachmentId = call.parameters["attachmentId"] ?: ""
            val (data, mimeType, filename) = attachmentService.getAttachmentFile(attachmentId)
            call.response.header(
                HttpHeaders.ContentDisposition,
                ContentDisposition.Inline.withParameter(ContentDisposition.Parameters.FileName, filename).toString()
            )
            call.respondBytes(data, ContentType.parse(mimeType))
        }

        // GET /api/vendors/{vendorId}/attachments/{attachmentId}/download - Download file
        get("/{attachmentId}/download") {
            call.authenticateUser()
            val attachmentId = call.parameters["attachmentId"] ?: ""
            val (data, mimeType, filename) = attachmentService.getAttachmentFile(attachmentId)
            call.response.header(
                HttpHeaders.ContentDisposition,
                ContentDisposition.Attachment.withParameter(ContentDisposition.Parameters.FileName, filename).toString()
            )
            call.respondBytes(data, ContentType.parse(mimeType))
        }

        // PUT /api/vendors/{vendorId}/attachments/{attachmentId} - Update metadata
        put("/{attachmentId}") {
            call.authenticateUser()
            val user = call.currentUser
            val attachmentId = call.parameters["attachmentId"] ?: ""
            val req = call.receive<UpdateVendorAttachmentReq>()
            val updated = attachmentService.updateAttachment(user.role, attachmentId, req.title, req.description)
            call.respond(HttpStatusCode.OK, successResponse(updated))
        }

        // DELETE /api/vendors/{vendorId}/attachments/{attachmentId} - Delete attachment
        delete("/{attachmentId}") {
            call.authenticateUser()
            val user = call.currentUser
            val attachmentId = call.parameters["attachmentId"] ?: ""
            attachmentService.deleteAttachment(user.role, attachmentId)
            call.respond(HttpStatusCode.OK, successResponse(mapOf("message" to "Attachment deleted successfully")))
        }
    }
}
